crates/red_knot_python_semantic/src/types.rs

```diff
@@ -721,6 +721,45 @@ impl<'db> Type<'db> {
             Type::Tuple(_) => builtins_symbol_ty(db, "tuple"),
         }
     }
+
+    /// Return the string representation of this type when converted to string as it would be
+    /// provided by the `__str__` method. If that can't be determined, return `None`.
+    ///
+    /// When not available, this should fall back to the value of `[Type::repr]`.
+    /// Note: this method is used in the builtins `format`, `print`, `str.format` and `f-strings`.
+    pub fn str(&self, db: &'db dyn Db) -> Option<Type<'db>> {
```

Comment:
Sorry, I wasn't clear; I don't think there's any reason for this to return an `Option`. The runtime enforces that `__str__` methods must return a string (otherwise it fails with a `TypeError`), so we can trust this. That means the only possible return values here are `str`, `LiteralString`, or a `StringLiteral`. So in any "can't be determined" case, we can always correctly fall back to returning the `str` type; there is no truly "unknown" case.

Reply:
Oh I didn't get that indeed, sorry! I'll fix it soon 🙂